File: src/ts/ssh-keys/sec1KeyFormatter.ts
//
//  Copyright (c) Microsoft Corporation. All rights reserved.
//

import {
	SshAlgorithms,
	KeyPair,
	BigInt,
	DerReader,
	DerWriter,
	DerType,
	ECDsa,
	ECParameters,
} from '@microsoft/dev-tunnels-ssh';
import { KeyFormatter, getKeyEncryptionAlgorithm } from './keyFormatter';
import { KeyData } from './keyData';
import { Pkcs1KeyFormatter } from './pkcs1KeyFormatter';

/** Provides import/export of the SEC1 key format. */
export class Sec1KeyFormatter implements KeyFormatter {
	private static readonly privateKeyType = 'EC PRIVATE KEY';

	public async import(keyData: KeyData): Promise<KeyPair | null> {
		if (!keyData) throw new TypeError('KeyData object expected.');

		if (keyData.keyType === Sec1KeyFormatter.privateKeyType || !keyData.keyType) {
			const parameters = Sec1KeyFormatter.importECPrivate(keyData.data);
			const keyPair = new ECDsa.KeyPair();
			await keyPair.importParameters(parameters);
			return keyPair;
		}

		return null;
	}

	public async export(keyPair: KeyPair, includePrivate: boolean): Promise<KeyData> {
		if (!keyPair) throw new TypeError('KeyPair object expected.');

		if (!includePrivate) {
			throw new Error('Public-only export is not supported by this format.');
		}

		if (keyPair instanceof ECDsa.KeyPair) {
			if (!keyPair.hasPublicKey) {
				throw new Error('KeyPair object does not have a public key.');
			} else if (!keyPair.hasPrivateKey) {
				throw new Error('KeyPair object does not have a private key.');
			}

			let keyData = new KeyData();
			const parameters = await keyPair.exportParameters();
			keyData.keyType = Sec1KeyFormatter.privateKeyType;
			keyData.data = Sec1KeyFormatter.exportECPrivate(parameters);

			return keyData;
		} else {
			throw new Error('KeyPair class not supported.');
		}
	}

	public async decrypt(keyData: KeyData, passphrase: string | null): Promise<KeyData | null> {
		if (!keyData) throw new TypeError('KeyData object expected.');

		if (keyData.keyType === Sec1KeyFormatter.privateKeyType || !keyData.keyType) {
			const procTypeHeader = keyData.headers.get('Proc-Type');
			if (procTypeHeader === '4,ENCRYPTED') {
				if (!passphrase) {
					throw new Error('A passphrase is required to decrypt the key.');
				}

				return Pkcs1KeyFormatter.decryptPrivate(keyData, passphrase);
			} else {
				return keyData;
			}
		}

		return null;
	}

	public async encrypt(keyData: KeyData, passphrase: string): Promise<KeyData> {
		if (!keyData) throw new TypeError('KeyData object expected.');

		if (keyData.keyType === Sec1KeyFormatter.privateKeyType) {
			throw new Error(
				'SEC1 export with passphrase is not supported because the format uses ' +
					'a weak key derivation algorithm. Use PKCS#8 to export a ' +
					'passphrase-protected private key.',
			);
		} else {
			throw new Error(`Unsupported key type: ${keyData.keyType}`);
		}
	}

	private static exportECPrivate(ec: ECParameters): Buffer {
		const curve = ECDsa.curves.find((c) => c.oid === ec.curve.oid)!;
		const keySizeInBytes = Math.ceil(curve.keySize / 8);

		const writer = new DerWriter(Buffer.alloc(1024));

		writer.writeInteger(BigInt.fromInt32(1)); // version
		writer.writeOctetString(ec.d!.toBytes({ unsigned: true, length: keySizeInBytes }));

		const curveWriter = new DerWriter(Buffer.alloc(100));
		curveWriter.writeObjectIdentifier(ec.curve.oid!);
		writer.writeTagged(0, curveWriter);

		const publicKeyWriter = new DerWriter(Buffer.alloc(512));
		const x = ec.x.toBytes({ unsigned: true, length: keySizeInBytes });
		const y = ec.y.toBytes({ unsigned: true, length: keySizeInBytes });
		const publicKeyData = Buffer.alloc(1 + x.length + y.length);
		publicKeyData[0] = 4; // Indicates uncompressed curve format
		x.copy(publicKeyData, 1);
		y.copy(publicKeyData, 1 + x.length);
		publicKeyWriter.writeBitString(publicKeyData);
		writer.writeTagged(1, publicKeyWriter);

		return writer.toBuffer();
	}

	private static importECPrivate(keyBytes: Buffer): ECParameters {
		const reader = new DerReader(keyBytes);
		const version = reader.readInteger().toInt32();
		if (version !== 1) {
			throw new Error(`Unsupported SEC1 format version: ${version}`);
		}

		const d = BigInt.fromBytes(reader.readOctetString(), { unsigned: true });

		const curveReader = reader.tryReadTagged(0);
		if (!curveReader) {
			throw new Error('SEC1 curve info not found.');
		}

		const curveOid = curveReader.readObjectIdentifier();
		const curveName = ECDsa.curves.find((c) => c.oid === curveOid)?.name;

		const publicKeyReader = reader.tryReadTagged(1);
		if (!publicKeyReader) {
			throw new Error('SEC1 public key data not found.');
		}

		const xy = publicKeyReader.readBitString();
		if (xy.length % 2 !== 1) {
			throw new Error(`Unexpected key data length: ${xy.length}`);
		}

		const x = BigInt.fromBytes(xy.slice(1, 1 + (xy.length - 1) / 2), { unsigned: true });
		const y = BigInt.fromBytes(xy.slice(1 + (xy.length - 1) / 2), { unsigned: true });
		const ec: ECParameters = {
			curve: { name: curveName, oid: curveOid },
			x,
			y,
			d,
		};
		return ec;
	}
}

File: src/cs/Ssh.Keys/Sec1KeyFormatter.cs
// Copyright (c) Microsoft. All rights reserved.

using System;
using Microsoft.DevTunnels.Ssh.Algorithms;
using Microsoft.DevTunnels.Ssh.IO;
using ECCurve = System.Security.Cryptography.ECCurve;
using ECParameters = System.Security.Cryptography.ECParameters;
using ECPoint = System.Security.Cryptography.ECPoint;

namespace Microsoft.DevTunnels.Ssh.Keys;

/// <summary>
/// Provides import/export of the <see cref="KeyFormat.Sec1" /> key format.
/// </summary>
public class Sec1KeyFormatter : IKeyFormatter
{
	private const string PrivateKeyType = "EC PRIVATE KEY";

	/// <inheritdoc/>
	public IKeyPair? Import(KeyData keyData)
	{
		if (keyData == null) throw new ArgumentNullException(nameof(keyData));

		ECParameters? parameters = null;
		if (keyData.KeyType == PrivateKeyType || string.IsNullOrEmpty(keyData.KeyType))
		{
			parameters = ImportECPrivate(keyData.Data);
		}

		if (parameters != null)
		{
			ECDsa.KeyPair? keyPair = new ECDsa.KeyPair();
			keyPair.ImportParameters(parameters.Value);
			return keyPair;
		}

		return null;
	}

	/// <inheritdoc/>
	public KeyData Export(IKeyPair keyPair, bool includePrivate)
	{
		if (keyPair == null) throw new ArgumentNullException(nameof(keyPair));

		if (!includePrivate)
		{
			throw new NotSupportedException("Public-only export is not supported by this format.");
		}

		if (keyPair is ECDsa.KeyPair ecKeyPair)
		{
			return ExportECPrivate(ecKeyPair);
		}
		else
		{
			throw new NotSupportedException(
				$"SEC1 format does not support key algorithm: {keyPair.KeyAlgorithmName}");
		}
	}

	/// <inheritdoc/>
	public KeyData? Decrypt(KeyData keyData, string? passphrase)
	{
		if (keyData == null) throw new ArgumentNullException(nameof(keyData));

		if (keyData.KeyType == PrivateKeyType || string.IsNullOrEmpty(keyData.KeyType))
		{
			if (keyData.Headers.TryGetValue("Proc-Type", out var procType) &&
				procType == "4,ENCRYPTED")
			{
				if (string.IsNullOrEmpty(passphrase))
				{
					throw new UnauthorizedAccessException(
						"A passphrase is required to decrypt the key.");
				}

				return Pkcs1KeyFormatter.DecryptPrivate(keyData, passphrase!);
			}
			else
			{
				return keyData;
			}
		}

		return null;
	}

	/// <inheritdoc/>
	public KeyData Encrypt(KeyData keyData, string passphrase)
	{
		if (keyData == null) throw new ArgumentNullException(nameof(keyData));

		if (keyData.KeyType == PrivateKeyType)
		{
			throw new NotSupportedException(
				"SEC1 export with passphrase is not supported because the format uses " +
				"a weak key derivation algorithm. Use a different format to export a " +
				"passphrase-protected private key.");
		}
		else
		{
			throw new InvalidOperationException($"Unexpected key type: {keyData.KeyType}");
		}
	}

	private static ECParameters ImportECPrivate(Buffer data)
	{
		var reader = new DerReader(data);

		var version = reader.ReadInteger().ToInt32();
		if (version != 1)
		{
			throw new NotSupportedException($"Unsupported SEC1 format version: {version}");
		}

		var d = reader.ReadOctetString().ToArray(); // Private key

		if (!reader.TryReadTagged(0, out var curveReader))
		{
			throw new InvalidOperationException("SEC1 curve info not found.");
		}

		var curveOid = curveReader.ReadObjectIdentifier();
		var curve = ECCurve.CreateFromOid(curveOid);

		if (!reader.TryReadTagged(1, out var publicKeyReader))
		{
			throw new InvalidOperationException("SEC1 public key data not found.");
		}

		var publicKeyData = publicKeyReader.ReadBitString();
		if (publicKeyData.Count % 2 != 1)
		{
			throw new InvalidOperationException(
				$"Unexpected key data length: {publicKeyData.Count}");
		}

		var dataFormat = publicKeyData[0];

		// 4 = uncompressed curve format
		if (dataFormat != 4)
		{
			throw new InvalidOperationException($"Unexpected curve format: {dataFormat}");
		}

		var keyLength = (publicKeyData.Count - 1) / 2;
		var x = publicKeyData.Slice(1, keyLength).ToArray();
		var y = publicKeyData.Slice(1 + keyLength, keyLength).ToArray();

		return new ECParameters
		{
			Curve = curve,
			Q = new ECPoint
			{
				X = x,
				Y = y,
			},
			D = d,
		};
	}

	private static KeyData ExportECPrivate(ECDsa.KeyPair keyPair)
	{
		var parameters = keyPair.ExportParameters(true);

		var writer = new DerWriter();

		writer.WriteInteger(BigInt.FromInt32(1)); // version

		writer.WriteOctetString(parameters.D!);

		var curveWriter = new DerWriter();
		curveWriter.WriteObjectIdentifier(parameters.Curve.Oid);
		writer.WriteTagged(0, curveWriter);

		var publicKeyWriter = new DerWriter();
		var publicKeyData = new Buffer(1 + parameters.Q.X!.Length + parameters.Q.Y!.Length);
		publicKeyData[0] = (byte)4; // Indicates uncompressed curve format
		Buffer.From(parameters.Q.X).CopyTo(publicKeyData, 1);
		Buffer.From(parameters.Q.Y).CopyTo(publicKeyData, 1 + parameters.Q.X.Length);
		publicKeyWriter.WriteBitString(publicKeyData);
		writer.WriteTagged(1, publicKeyWriter);

		var data = writer.ToArray();

		return new KeyData
		{
			KeyType = PrivateKeyType,
			Data = data,
		};
	}
}
